File: Tic-Tac-Toe.py
from random import randint
from IPython.display import clear_output

# Display board
def display_board(board):
    clear_output()
    print()
    print('\t\t ' + board[7] + ' | ' + board[8] + ' | ' + board[9])
    print('\t\t---|---|---')
    print('\t\t ' + board[4] + ' | ' + board[5] + ' | ' + board[6])
    print('\t\t---|---|---')
    print('\t\t ' + board[1] + ' | ' + board[2] + ' | ' + board[3])
    print()
    
# Choose the player to go first
def go_first():
    first = randint(1,2)
    if first == 1:
        return 'Player 1'
    else:
        return 'Player 2'

# Prompt ask players to pick their marker
def choose_marker():
    while True:
        marker = input('Pick your mark (X/O): ').upper()
        if marker == 'X':
            return ('X', 'O')
        elif marker == 'O':
            return ('O', 'X')
        else:
            print("INVALID INPUT!!!\nPlease pick again.\n")

# Check if the position is good or not
def good_position(board, position):
    return board[position] == ' '

# Prompt ask players to pick the position
def position_input(board):
    position = 0
    while position not in range(1,10) or not good_position(board, position):
        position = int(input('Choose a number to place your mark (1 - 9): '))
        if position not in range(1,10):
            print('INVALID!!! The input is out of acceptable range (1 - 9)')
        elif not good_position(board, position):
            print('That position has already occupied!!! Choose another place.')
    return position

# Place the marker on the board
def place_marker(board, position, marker):
    board[position] = marker

# The game is tie if the board is full
def full_board(board):
    for n in range(1,10):
        if good_position(board, n):
            return False
    return True

# Win condition of the game
def win_check(board, marker):
    return ((board[1] == board[2] == board[3] == marker) or
            (board[4] == board[5] == board[6] == marker) or
            (board[7] == board[8] == board[9] == marker) or
            (board[1] == board[4] == board[7] == marker) or
            (board[2] == board[5] == board[8] == marker) or
            (board[3] == board[6] == board[9] == marker) or
            (board[1] == board[5] == board[9] == marker) or
            (board[3] == board[5] == board[7] == marker))

# Replay the game
def replay():
    choice = input('Want to play again? Yes (y) or No (n): ').lower()
    return choice == 'yes' or choice == 'y'

# Start the game
def game_start():
    print('\t WELCOME TO TIC-TAC-TOE!!!')
    instruction_board = ['', '1', '2', '3', '4', '5', '6', '7', '8', '9']
    display_board(instruction_board)
    print("--- Use the numbers as instructions to place your marks ---\n")

    while True:
        board = [' '] * 10
        play = input("Ready to play? Yes (y) or No (n): ").lower()
        if play == 'yes' or play == 'y':
            game_start = True
            # Set up players and their markers
            player1_marker, player2_marker = choose_marker()
            player = go_first()
            print(player + ' will go first!')
        else:
            game_start = False

        while game_start:
            if player == 'Player 1':
                display_board(board)
                print(player1_marker + "'s turn")
                # Choose a position
                position = position_input(board)
                # Place the marker on this position
                place_marker(board, position, player1_marker)
                
                if win_check(board, player1_marker):
                    display_board(board)
                    print('Player 1 is the winner!')
                    game_start = False
                elif full_board(board):
                    display_board(board)
                    print('Good game! It is a tie')
                    game_start = False
                else:
                    player = 'Player 2'

            else:
                display_board(board)
                print(player2_marker + "'s turn")
                # Choose a position
                position = position_input(board)
                # Place the marker on this position
                place_marker(board, position, player2_marker)
                
                if win_check(board, player2_marker):
                    display_board(board)
                    print('Player 2 is the winner!')
                    game_start = False
                elif full_board(board):
                    display_board(board)
                    print('Good game! It is a tie')
                    game_start = False
                else:
                    player = 'Player 1'
    
        if not replay():
            break

game_start()
       
if __name__ == "__main__":
    pass

File: Tic-Tac-Toe.cpp
#include <iostream>
#include <vector>
#include <string>

using std::cout;
using std::cin;
using std::endl;
using std::vector;

int player = 1;
char turn = 'X';
int choice;
bool draw = false;
int i, j;

// give instruction to players
void instruction();

// display the board
void display_board(std::vector<std::vector<char>>& board);

// check if the game have a winner, or game draw
bool game_over(std::vector<std::vector<char>>& board);

// get player's input and update the board
void player_input(std::vector<std::vector<char>>& board);

// annouce the winner
void winner(std::vector<std::vector<char>>& board);

int main()
{
	std::vector<std::vector<char>> board = { { '1', '2', '3'}, {'4', '5', '6' }, {'7', '8', '9'} };
	instruction();
	display_board(board);
	while (!game_over(board))
	{
		player_input(board);
		display_board(board);
	}
	winner(board);
}


void instruction()
{
	cout << "\t\t\tWelcome to the Tic-Tac-Toe game!\n\n";
	cout << "\tTwo players will make their move by entering a number from 1-9.\n";
	cout << "As the board display, the number will correspond to the position on the board.\n\n";
}


void display_board(std::vector<std::vector<char>>& board)
{
	cout << "-------------" << endl;
	for (auto vec : board)
	{
		for (char c : vec)
		{
			cout << "| " << c << " ";
		}
		cout << "|" << endl;
		cout << "-------------" << endl;
	}
}


bool game_over(std::vector<std::vector<char>>& board)
{
	if (board[0][0] == board[0][1] && board[0][1] == board[0][2])
		return true;
	else if (board[1][0] == board[1][1] && board[1][1] == board[1][2])
		return true;
	else if (board[2][0] == board[2][1] && board[2][1] == board[2][2])
		return true;
	else if (board[0][0] == board[1][0] && board[1][0] == board[2][0])
		return true;
	else if (board[0][1] == board[1][1] && board[1][1] == board[2][1])
		return true;
	else if (board[0][2] == board[1][2] && board[1][2] == board[2][2])
		return true;
	else if (board[0][0] == board[1][1] && board[1][1] == board[2][2])
		return true;
	else if (board[0][2] == board[1][1] && board[1][1] == board[2][0])
		return true;
	else if (board[0][0] != '1' && board[0][1] != '2' && board[0][2] != '3' &&
		board[1][0] != '4' && board[1][1] != '5' && board[1][2] != '6' &&
		board[2][0] != '7' && board[2][1] != '8' && board[2][2] != '9')
	{
		draw = true;
		return true;
	}
	return false;
}


void player_input(std::vector<std::vector<char>>& board)
{
	if (turn == 'X') {
		cout << "Player 1 - [X] turn - ";
	}
	else if (turn == 'O') {
		cout << "Player 2 - [O] turn - ";
	}
	cout << "Player choose position: ";
	while (!(cin >> choice)) {
		cout << "Wrong input! Please enter your choice again: ";
		cin.clear();
		cin.ignore();
	}
	// switch case to get which i and j will be update
	switch (choice) {
	case 1: i = 0; j = 0; break;
	case 2: i = 0; j = 1; break;
	case 3: i = 0; j = 2; break;
	case 4: i = 1; j = 0; break;
	case 5: i = 1; j = 1; break;
	case 6: i = 1; j = 2; break;
	case 7: i = 2; j = 0; break;
	case 8: i = 2; j = 1; break;
	case 9: i = 2; j = 2; break;
	}

	if (turn == 'X' && board[i][j] != 'X' && board[i][j] != 'O')
	{
		// updating the position for 'X' symbol if
		// it is not already occupied
		board[i][j] = 'X';
		turn = 'O';
	}
	else if (turn == 'O' && board[i][j] != 'X' && board[i][j] != 'O')
	{
		// updating the position for 'O' symbol if
		// it is not already occupied
		board[i][j] = 'O';
		turn = 'X';
	}
	else
	{
		// if input position already filled
		cout << "\nPlease choose another!!\n\n";
		player_input(board);
	}
}


void winner(std::vector<std::vector<char>>& board)
{
	if (turn == 'X' && draw == false)
		cout << "\n-----Congratulations! Player with 'O' is the winner-----";
	else if (turn == 'O' && draw == false)
		cout << "\n-----Congratulations! Player with 'X' is the winner-----";
	else
		cout << "\n-----Both players are tie-----!!!\n";
}